src/main/java/de/tum/in/www1/artemis/service/GradingScaleService.java

```diff
@@ -0,0 +1,118 @@
+package de.tum.in.www1.artemis.service;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.domain.GradeStep;
+import de.tum.in.www1.artemis.domain.GradingScale;
+import de.tum.in.www1.artemis.repository.GradeStepRepository;
+import de.tum.in.www1.artemis.repository.GradingScaleRepository;
+import de.tum.in.www1.artemis.web.rest.errors.BadRequestAlertException;
+import de.tum.in.www1.artemis.web.rest.errors.EntityNotFoundException;
+
+@Service
+public class GradingScaleService {
+
+    private final GradingScaleRepository gradingScaleRepository;
+
+    private final GradeStepRepository gradeStepRepository;
+
+    public GradingScaleService(GradingScaleRepository gradingScaleRepository, GradeStepRepository gradeStepRepository) {
+        this.gradeStepRepository = gradeStepRepository;
+        this.gradingScaleRepository = gradingScaleRepository;
+    }
+
+    /**
+     * Maps a grade percentage to a valid grade step within the grading scale or throws an exception if no match was found
+     *
+     * @param percentage the grade percentage to be mapped
+     * @param gradingScaleId the identifier for the grading scale
+     * @return grade step corresponding to the given percentage
+     */
+    public GradeStep matchPercentageToGradeStep(double percentage, Long gradingScaleId) {
+        if (percentage < 0 || percentage > 100) {
+            throw new BadRequestAlertException("Grade percentages must be between 0 and 100", "gradeStep", "invalidGradePercentage");
+        }
+        List<GradeStep> gradeSteps = gradeStepRepository.findByGradingScale_Id(gradingScaleId);
+        Optional<GradeStep> matchingGradeStep = gradeSteps.stream().filter(gradeStep -> gradeStep.matchingGradePercentage(percentage)).findFirst();
+        if (matchingGradeStep.isPresent()) {
+            return matchingGradeStep.get();
+        }
+        else {
+            throw new EntityNotFoundException("No grade step in selected grading scale matches given percentage");
```

Comment:
Two things:
- I am not sure that this is the right exception to throw here, as this exception is normally used in our codebase to signify that something does not exist in the database. 
- I think throwing an exception here is weird (not wrong), maybe it would be better to return an `Optional<GradeStep>` ? **What is the opinion of the others?**

Reply:
I theory this exception should never be thrown, because of two things:
- First of all the server can already handle any invalid grading scale, so whenever a grading scale exists for this particular course (or exam), it would contain mappings for all grade percentages between 0 and 100, if it doesn't the resource handles this case and returns a 404 Not Found
- All invalid percentages (<0, >100) get handled in the first line of this method

Therefore, this else statement is more like a fail-safe, if for some reason the mapping could **not** be **found** in the database. Bad request here makes no sense because the route has to have been correct in order to have entered this method, and the percentage was within the permitted boundries.